privacyidea/lib/token.py

```diff
@@ -2597,3 +2598,80 @@ def list_tokengroups(tokengroup=None):
 
     return tgs
+
+
+def token_dump(token, tokenowner=True):
+    """
+    Store the database columns of the token into a dict.
+    Also store the tokeninfo into a list of dicts.
+
+    :param token: A token object
+    :param tokenowner: Also dump the tokenowners
+    :type tokenowner: bool
+    :return: a dict, containing the token and the tokeninfo
+    """
+    token_dict = token._to_dict()
+    if tokenowner:
+        # handle all assigned users
+        owners = []
+        for owner in token.owners:
+            owners.append({"uid": owner.uid,
+                           "login": owner.login,
+                           "resolver": owner.resolver,
+                           "realm": owner.realm})
+        token_dict["owners"] = owners
+    return token_dict
+
+
+def token_load(token_dict, tokenowner=True, overwrite=False):
+    """
+    Load the token that has previously been dumped with the function token_dump.
+
+    :param token_dict: The token in a dict
+    :param tokenowner: The tokenowner should also be assigned. If the tokenowner can not be found or
+        identified, an exception is raised.
+    :type tokenowner: bool
+    :param overwrite: If a token with the given serial number already exist, it should be overwritten. If the token
+        should not be overwritten but already exists, an exception is raised.
+    :type overwrite: bool
+    :return:
+    """
+    serial = token_dict.get("serial")
+    old_token_obj = get_one_token(serial=serial, silent_fail=True)
+    # Check if overwriting is not allowed and the token already exists
+    if not overwrite and old_token_obj:
+        raise TokenAdminError("Token already exists!")
+
+    tokeninfos = token_dict.get("info_list", {})
+    hashed_pin = token_dict.get("_hashed_pin")  # Uncomment if needed
+    # Creating a new dictionary without special keys
+    stripped_token = {k: v for k, v in token_dict.items() if k not in ["info_list", "owners", "_hashed_pin"]}
+    # Initialize or update token
+    token = init_token(stripped_token)
+    # Add token information
+    for key, value in tokeninfos.items():
+        token.add_tokeninfo(key, value)
+    # Set PIN if hashed_pin is available
+    if hashed_pin:
+        token.token.pin_hash = hashed_pin
+        token.token.save()
+
+    # Add the tokenowners
+    if tokenowner:
+        for owner in token_dict.get("owners", []):
+            # uid, login, resolver, realm
+            # We might need to create the user object
+            # Now we add the user as tokenowner
+            try:
+                loginname = get_username(owner.get("uid"), owner.get("resolver"))
+                if loginname == owner.get("login"):
+                    u = User(login=owner.get("login"),
+                             resolver=owner.get("resolver"),
+                             realm=owner.get("realm"),
+                             uid=owner.get("uid"))
+                    token.add_user(u)
+            except Exception as ex:
```

Comment:
My intention was to try to assign the user but issue a log output, if the user can not be found.
I is by design, that no exception is raised, so that a couple of non existing users will not break the complete import!

Reply:
Ah, ok. Will adapt.

Reply:
Done in 4e3bf3ade